### PR TITLE
Pass context with environment to backingViewDescription

### PR DESCRIPTION
**Pull request by kylebshr**
Adds an `ElementContext` object for passing information to `backingViewDescription`, and adds the environment to the context (along with the existing `bounds` and `subtreeExtent`). This allows view-backed elements to access the environment they're rendered in.

### CODE REVIEW

File: BlueprintUI/Sources/Element/Element.swift [hunk 1]
```diff
@@ -61,6 +61,14 @@ public protocol Element {
     ///                            `subtreeExtent` will be nil if there are no children.
     ///
     /// - Returns: An optional `ViewDescription`.
-    func backingViewDescription(bounds: CGRect, subtreeExtent: CGRect?) -> ViewDescription?
+    func backingViewDescription(bounds: CGRect, subtreeExtent: CGRect?, with context: ElementContext) -> ViewDescription?
```
Comment (kyleve) [inline BlueprintUI/Sources/Element/Element.swift:64]: I assume just adding the context here so refactoring is easier? Would be nice to make this one param...

Comment (kyleve) [reply BlueprintUI/Sources/Element/Element.swift:64]: (Looks like there's 134 of these in SPOS)

```
~/Desktop/Development/Register1(kve/bump-listable-jun-15-2021):0 git grep "func backingViewDescription(bounds:" | wc -l
     134
```

File: BlueprintUI/Sources/Element/ElementContent.swift [hunk 2]
```diff
@@ -413,6 +415,7 @@ private struct LazyStorage: ContentStorage {
         let node = LayoutResultNode(
             element: child,
             layoutAttributes: childAttributes,
+            environment: environment,
```
Comment (kyleve) [inline BlueprintUI/Sources/Element/ElementContent.swift:418]: Oh smart! I guess adding this to the layout result does get us the right environment; nice!!

File: BlueprintUI/Sources/Element/Element.swift (hunk 1 above)
Comment (kyleve) [reply BlueprintUI/Sources/Element/Element.swift:64]: (Looks like about 120 of these return views; the rest return nil.)

File: BlueprintUI/Sources/BlueprintView/BlueprintView.swift [hunk 3]
```diff
@@ -36,6 +36,13 @@ public final class BlueprintView: UIView {
 
     private let rootController: NativeViewController
 
+    public var environment: Environment {
```
Comment (kyleve) [inline BlueprintUI/Sources/BlueprintView/BlueprintView.swift:39]: Ah ha, yay! I'd been wanting this for a while.

File: BlueprintUI/Sources/Element/ElementContent.swift (hunk 2 above)
Comment (kylebshr) [reply BlueprintUI/Sources/Element/ElementContent.swift:418]: Less "smart" and more me just adding the environment to the `backingViewDescription` and finding myself here, where there was already an environment passed in 😂 but yeah, glad it worked it out like that!

File: BlueprintUI/Sources/BlueprintView/BlueprintView.swift (hunk 3 above)
Comment (kylebshr) [reply BlueprintUI/Sources/BlueprintView/BlueprintView.swift:39]: Yeah me too - added a test as well

**Review by n8chur — APPROVED** [review]

File: BlueprintUI/Sources/Element/Element.swift [hunk 4]
```diff
@@ -56,11 +56,29 @@ public protocol Element {
...
     /// Elements become view-backed by returning a `ViewDescription` value from this method.
     ///
-    /// - Parameter bounds: The bounds of this element after layout is complete.
-    /// - Parameter subtreeExtent: A rectangle in the local coordinate space that contains any children.
-    ///                            `subtreeExtent` will be nil if there are no children.
+    /// - Parameter context: The context this element is rendered in.
     ///
     /// - Returns: An optional `ViewDescription`.
-    func backingViewDescription(bounds: CGRect, subtreeExtent: CGRect?) -> ViewDescription?
+    func backingViewDescription(in context: ElementContext) -> ViewDescription?
```
Comment (kyleve) [inline BlueprintUI/Sources/Element/Element.swift:62]: You're going to hate me (sorry); what do we think about:

```
func viewDescription(with context: ElementContext) -> ViewDescription?
```

Also, do we want `ElementContext`; or should it be `ViewDescriptionContext` or something?

Comment (kylebshr) [reply BlueprintUI/Sources/Element/Element.swift:62]: Both changes are fine with me! I had `with` originally but `in` seemed kinda nice with the one param? idk, fine with `with`

Comment (kyleve) [reply BlueprintUI/Sources/Element/Element.swift:62]: (I think `ElementContext` is fine; just wondering if we'll ever want that name for something else)

Comment (kylebshr) [reply BlueprintUI/Sources/Element/Element.swift]: Already did `ViewDescriptionContext`

**Review by kyleve — APPROVED** [review]

**Pull request merged**